### PR TITLE
Chapter 9: Advanced Login

### PR DESCRIPTION
**Pull request by ainulfarizah**
ready

### CODE REVIEW

**Comment by ainulfarizah** [comment]
ready

File: app/assets/stylesheets/sessions.scss [hunk 1]
```diff
@@ -0,0 +1,3 @@
+// Place all the styles related to the Sessions controller here.
```
Comment (ghost) [inline app/assets/stylesheets/sessions.scss]: delete this line please

File: app/controllers/sessions_controller.rb [hunk 2]
```diff
@@ -2,11 +2,20 @@ class SessionsController < ApplicationController
...
       log_in @user
       redirect_to @user
+=======
+    user = User.find_by email: params[:session][:email].downcase
+    if @user && @user.authenticate params[:session][:password]
+      # Log the user in and redirect to the user's show page.
+      log_in user
+      params[:session][:remember_me] == "1" ? remember(@user) : forget(@user)
+      redirect_to user
+>>>>>>> Chapter 9
```
Comment (ghost) [inline app/controllers/sessions_controller.rb]: ???? what is that?

File: app/helpers/sessions_helper.rb [hunk 3]
```diff
@@ -42,9 +50,17 @@ def remember user
   end
 
   # Forgets a persistent session.
+<<<<<<< HEAD
```
Comment (ghost) [inline app/helpers/sessions_helper.rb]: ???/

**Review by ghost — COMMENTED** [review]
please fix chose your code you want to save

**Comment by ainulfarizah** [comment]
ready

File: config/routes.rb [hunk 4]
```diff
@@ -1,4 +1,7 @@
 Rails.application.routes.draw do
+  get 'sessions/new'
```
Comment (ngocvt-0484) [inline config/routes.rb]: line 2,3,4 delete please
you defined at line 13, 14,15

File: test/controllers/sessions_controller_test.rb [hunk 5]
```diff
@@ -0,0 +1,18 @@
+require "test_helper"
+
+class SessionsControllerTest < ActionDispatch::IntegrationTest
+  test "should get new" do
+    get sessions_new_url
+    assert_response :success
+  end
```
Comment (ngocvt-0484) [inline test/controllers/sessions_controller_test.rb]: dont need do test
 so please delete this file

**Comment by ngocvt-0484** [comment]
delete app/assets/stylesheets/sessions.scss
because it not use

**Comment by ainulfarizah** [comment]
ready

File: app/models/user.rb [hunk 6]
```diff
@@ -1,9 +1,11 @@
 class User < ApplicationRecord
+  attr_accessor :remember_token
   before_save :downcase_email
 
   validates :name, presence: true, length: {minimum: 6}
   VALID_EMAIL_REGEX = /\A[\w+\-.]+@[a-z\d\-.]+\.[a-z]+\z/i
   validates :email, presence: true, length: {maximum: Settings.user.email.max_length}, format: {with: VALID_EMAIL_REGEX}
+  format: {with: VALID_EMAIL_REGEX}
```
Comment (huongnt-2545) [inline app/models/user.rb]: remove this line

File: app/models/user.rb [hunk 7]
```diff
@@ -22,4 +24,34 @@ def User.digest string
   def downcase_email
```
Comment (huongnt-2545) [inline app/models/user.rb]: move all methods have private scope place to the end of class

**Comment by ainulfarizah** [comment]
ready

File: app/models/user.rb [hunk 8]
```diff
@@ -19,7 +20,37 @@ def User.digest string
 
   private
```
Comment (huongnt-2545) [inline app/models/user.rb]:
```suggestion
  class << self
    def new_token
      SecureRandom.urlsafe_base64
    end

    # Returns the hash digest of the given string.
    def digest string
      cost = if ActiveModel::SecurePassword.min_cost
               BCrypt::Engine::MIN_COST
             else
               BCrypt::Engine.cost
             end
      BCrypt::Password.create string, cost: cost
    end
  end
  def remember
    self.remember_token = User.new_token
    update_attribute :remember_digest, User.digest(remember_token)
  end

  def authenticated? remember_token
    BCrypt::Password.new(remember_digest).is_password? remember_token
  end

  # Forgets a user.
  def forget
    update_attribute :remember_digest, nil
  end
  
  private
  
  def downcase_email
    self.email.downcase!
  end
```

File: app/views/sessions/new.html.erb [hunk 9]
```diff
@@ -7,8 +8,12 @@
         <%= f.email_field :email, class: "form-control" %>
         <%= f.label :password %>
         <%= f.password_field :password, class: "form-control" %>
-        <%= f.submit t(:log_in), class: "btn btn-primary" %>
+        <%= f.label :remember_me, class: "checkbox inline" do %>
+          <%= f.check_box :remember_me %>
+          <span>Remember me on this computer</span>
       <% end %>
```
Comment (huongnt-2545) [inline app/views/sessions/new.html.erb]: indent

**Comment by ainulfarizah** [comment]
ready

**Review by huongnt-2545 — APPROVED** [review]

**Pull request merged**